zproject/backends.py

```diff
@@ -742,19 +742,13 @@ def sync_custom_profile_fields_from_ldap(
             var_name = "_".join(field.name.lower().split(" "))
             fields_by_var_name[var_name] = field
 
-        existing_values = {}
-        for data in user_profile.profile_data:
-            var_name = "_".join(data["name"].lower().split(" "))
-            existing_values[var_name] = data["value"]
-
         profile_data: List[Dict[str, Union[int, str, List[int]]]] = []
         for var_name, value in values_by_var_name.items():
             try:
                 field = fields_by_var_name[var_name]
             except KeyError:
                 raise ZulipLDAPException(f"Custom profile field with name {var_name} not found.")
-            if existing_values.get(var_name) == value:
-                continue
+
```

Comment:
This may be redundant, but it is a significant performance optimization?  The `do_update_user_custom_profile_data_if_changed` function does a bunch of get_or_create operations inside a transaction, so if we can skip calling it 1x per (field x user) in an organization, that's going to avoid potentially 10K database queries in a loop, right?  

As it is, we just do the one bulk query up front and then only do additional queries if anything has changed.

(If you agree with my thinking here, we should add comments explaining why we do the `existing_values` fetch, since I agree it's not otherwise obvious why this is important).

Reply:
@timabbott The old code was still running an extra query. The cost of figuring out "existing values" is needing to access `.profile_data`, which runs a query:
```
    @property
    def profile_data(self) -> ProfileData:
        values = CustomProfileFieldValue.objects.filter(user_profile=self)
        user_data = {
            v.field_id: {"value": v.value, "rendered_value": v.rendered_value} for v in values
        }
        data: ProfileData = []
        for field in custom_profile_fields_for_realm(self.realm_id):
            field_values = user_data.get(field.id, None)
            if field_values:
                value, rendered_value = field_values.get("value"), field_values.get(
                    "rendered_value"
                )
            else:
                value, rendered_value = None, None
            field_type = field.field_type
            if value is not None:
                converter = field.FIELD_CONVERTERS[field_type]
                value = converter(value)

            field_data = field.as_dict()
            data.append(
                {
                    "id": field_data["id"],
                    "name": field_data["name"],
                    "type": field_data["type"],
                    "hint": field_data["hint"],
                    "field_data": field_data["field_data"],
                    "order": field_data["order"],
                    "value": value,
                    "rendered_value": rendered_value,
                }
            )

        return data
```

So that actually adds a query per each user. But for each custom field that is *not* being updated (because it already has the target value) it does save a `get_or_create` query in `do_update_user_custom_profile_data_if_changed`. So overall I think you're right that it will reduce the number of queries - because usually most, or all, the fields will already be at the target values.

But I think the right approach might be to get rid of this code and improve `do_update_user_custom_profile_data_if_changed` to do the job better - it can access `user_profile.profile_data` to run the one bulk query and use that  to avoid redundant `get_or_create` calls?

Reply:
I think ultimately we probably just want to do a bulk query in the sync program for all CustomProfileFieldValues in the realm, and to do O(1) total queries.  

We should as a follow-up change `profile_data` to not be a property and instead a function; it's bad for readability of "what is the scalability of this code?" to have a property that actually does database queries behind the scenes, and we've tried to eliminate those elsewhere.